rfc/src/rfcs/0008-function-contracts.md

```diff
@@ -0,0 +1,580 @@
+
+   Next Kani must make sure that the approximation we specified actually holds.
+   This is in opposition to the related "stubbing" mechanism, where the
+   approximation is not checked but always trusted.
+
+   The contract must always overapproximate the function to guarantee soundness.
+   Specifically the domain (inputs) of the function described (by the `requires`
+   clause) must be *at most* as large as the input space for which the
+   function's behavior is well defined, and the codomain (outputs, described by
+   `ensures`) must be *at least* as large as the actual space of outputs the
+   function may produce.
+
+   For example in our case it would be permissible to use
+   `#[kani::requires(divisor > 100)]` (smaller permissible input domain) or
+   `#[kani::ensures(result < dividend + divisor)]` (larger possible output
+   domain), but `#[kani::ensures(result < dividend)]` (small output) is not
+   allowed.
+
+   To facilitate the check Kani needs a suitable environment to verify the
+   function in. For this proposal the environment must be provided by us (the
+   users). See [future possibilities](#future-possibilities) for a discussion
+   about the arising soundness issues and their remedies.
+
+   We provide the checking environment for our contract with a special new
+   `proof_for_contract` harness.
+
+   ```rs
+   #[kani::proof_for_contract(my_div)]
+   fn my_div_harness() {
+     my_div(kani::any(), kani::any())
+   }
+   ```
+
+   Similar to a unit-test harness for any other function, we are supposed to
+   create inputs for the function that are as generic/abstract as possible to
+   make sure we catch all edge cases, then call the function at least once with
+   those abstract inputs. If we forget to call `my_div` Kani would report an
+   error.
+   
+   Unlike a unit-test we can however elide any checks of the output and
+   post-call state. Instead Kani uses the conditions we specified in the
+   contract as checks. Preconditions (`requires`) are inserted as `kani::assume`
+   *before* the call to `my_div`, limiting the generic inputs to those the
+   function is actually defined for. Postconditions (`ensures`) are inserted as
+   `kani::assert` checks *after* the call to `my_div`.
+
+   The expanded version of our harness and function is equivalent to the following:
+
+   ```rs
+   #[kani::proof]
+   fn my_div_harness() {
+     let dividend = kani::any();
+     let divisor = kani::any();
+     kani::assume(divisor != 0);
+     let result = my_div(dividend, divisor);
+     kani::assert(result <= dividend);
+   }
+   ```
+
+   This expanded harness is then verified like any other harness but also gives
+   the green light for the next step: verified stubbing.
+
+3. In the last phase the **verified** contract is ready for us to use to
+   **stub** other harnesses.
+
+   Unlike in regular stubbing Kani there to be at least one associated
+   `proof_for_contract` harness for each function to stub *and* it requires all
```

Comment:
Are we actually planning on enforcing this? What about dependencies?

Reply:
I worry that contracts are easily used unsoundly (e.g. unchecked). Maybe an opt-out mechanism is the way to go. I think in the end this should be fixed with a cache that is discussed later